File: 압축/java/20230106/src/Solution.java
import java.util.*;

public class Solution {
    public ArrayList<Integer> solution(String msg) {
        ArrayList<Integer> answer = new ArrayList<>();
        HashMap<String, Integer> dictionary = new HashMap<>();

        int index = 26;

        initDictionary(dictionary);

        int size = msg.length();

        for (int i = 0; i < size; i += 1) {
            int a = 1;

            while (i + a <= size && dictionary.containsKey(msg.substring(i, i + a))) {
                a += 1;
            }

            if (i + a > size) {
                answer.add(dictionary.get(msg.substring(i)));

                break;
            }

            answer.add(dictionary.get(msg.substring(i, i + a - 1)));

            dictionary.put(msg.substring(i, i + a), index += 1);

            if (a > 1) {
                i += a - 2;
            }
        }

        return answer;
    }

    public void initDictionary(HashMap<String, Integer> dictionary) {
        char character = 'A';

        for (int i = 1; i < 27; i += 1) {
            dictionary.put(Character.toString(character), i);

            character += 1;
        }
    }
}

File: 압축/javascript/20230106/solution.js
function solution(msg) {
  const answer = compress(msg);
  return answer;
}

function compress(msg) {
  let message = msg;
  const answer = [];

  const dictionary = {};
  for (let i = 0; i < 26; i += 1) {
    dictionary[String.fromCharCode('A'.charCodeAt(0) + i)] = i + 1;
  }

  while (message.length > 0) {
    const length = wLength(message, dictionary);
    const w = message.substring(0, length);

    answer.push(dictionary[w]);
    message = message.substring(length);
    
    if (message.length > 0) {
      dictionary[w + message.substring(0, 1)] = Object.keys(dictionary).length + 1;
    }
  }

  return answer;
}

function wLength(msg, dictionary) {
  if (dictionary[msg]) {
    return msg.length;
  }

  for (let i = 0; i < msg.length; i += 1) {
    if (!dictionary[msg.substring(0, i + 1)]) {
      return i;
    }
  }

  return -1;
}
